File: e72/bac/macro/transmittance.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit

# =========================
# Input
# =========================
fname = "ag22k001.txt"
t_mm = 12.4                 # sample thickness [mm]
lam_eval_nm = 400.0         # wavelength for A_T

# thickness for transmission length
d_cm = t_mm / 10.0          # 10 mm = 1 cm

# =========================
# Load data
# =========================
data = np.loadtxt(fname)
lam_nm = data[:, 0].astype(float)
T = data[:, 1].astype(float) / 100.0   # transmittance (fraction)

# =========================
# Rayleigh model
# T = A * exp(-B / lambda^4),  B = C*t
# =========================
def model(lam_nm, A, B):
    return A * np.exp(-B / (lam_nm**4))

mask = (T > 0) & (T < 1.0)

A0 = np.max(T)
B0 = 1e10

popt, pcov = curve_fit(
    model,
    lam_nm[mask],
    T[mask],
    p0=[A0, B0],
    maxfev=20000
)

A_fit, B_fit = popt
A_err, B_err = np.sqrt(np.diag(pcov))

# =========================
# Rayleigh coefficient C
# =========================
C_nm4_per_mm = B_fit / t_mm
C_nm4_per_mm_err = B_err / t_mm

# unit conversion: nm^4/mm -> um^4/cm
C_um4_per_cm = C_nm4_per_mm * 1e-11
C_um4_per_cm_err = C_nm4_per_mm_err * 1e-11

# =========================
# Transmittance at 400 nm
# =========================
T400 = model(lam_eval_nm, A_fit, B_fit)

# =========================
# Transmission length A_T(400)
# A_T = -d / ln T
# =========================
if 0.0 < T400 < 1.0:
    AT400_cm = -d_cm / np.log(T400)
else:
    AT400_cm = np.nan

# =========================
# Print results (paper-ready)
# =========================
print("==== Rayleigh scattering fit (paper-ready) ====")
print(f"A = {A_fit:.6g} ± {A_err:.2g}")
print(f"C = ({C_um4_per_cm:.4g} ± {C_um4_per_cm_err:.2g})  μm^4 / cm")
print()
print(f"T({lam_eval_nm:.0f} nm) = {T400:.6f}  ({T400*100:.3f} %)")
print(f"A_T({lam_eval_nm:.0f} nm) = {AT400_cm:.6g} cm")
print(f"(using thickness d = {d_cm:.3f} cm = {t_mm} mm)")

# =========================
# Plot
# =========================
lam_dense = np.linspace(lam_nm.min(), lam_nm.max(), 2000)

plt.figure(figsize=(7.5, 4.5))
plt.plot(lam_nm, T, "o", ms=4, label="Data")
plt.plot(
    lam_dense,
    model(lam_dense, A_fit, B_fit),
    "-",
    lw=2,
    label=r"Fit: $A\,\exp(-Ct/\lambda^4)$"
)


plt.xlabel("Wavelength [nm]",fontsize=16)
plt.ylabel("Transmittance",fontsize=16)
plt.grid(alpha=0.3)
#plt.legend()
plt.tight_layout()
plt.savefig("ag22k001_rayleigh_fit.png", dpi=300, bbox_inches="tight")
plt.savefig("ag22k001_rayleigh_fit.pdf", bbox_inches="tight")
plt.show()

File: e72/bac/macro/transmittance.cc
#include <TCanvas.h>
#include <TGraph.h>
#include <TF1.h>
#include <TStyle.h>
#include <TAxis.h>
#include <TLegend.h>
#include <TMath.h>
#include <TFitResultPtr.h>
#include <TFitResult.h>

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cmath>

void transmittance()
{
  // =========================
  // Input
  // =========================
  std::string fname = "ag22k001.txt";
  double t_mm = 12.4;        // sample thickness [mm]
  double lam_eval_nm = 400.; // wavelength for A_T

  // thickness for transmission length
  double d_cm = t_mm / 10.0; // 10 mm = 1 cm

  // =========================
  // Load data
  // =========================
  std::ifstream fin(fname);
  if (!fin.is_open()) {
    std::cerr << "Error: cannot open file " << fname << std::endl;
    return;
  }

  std::vector<double> lam_nm;
  std::vector<double> T;

  double lam, trans_percent;
  while (fin >> lam >> trans_percent) {
    double trans = trans_percent / 100.0; // convert % -> fraction
    if (trans > 0.0 && trans < 1.0) {
      lam_nm.push_back(lam);
      T.push_back(trans);
    }
  }
  fin.close();

  if (lam_nm.empty()) {
    std::cerr << "Error: no valid data loaded from " << fname << std::endl;
    return;
  }

  // =========================
  // TGraph
  // =========================
  TGraph* g = new TGraph(lam_nm.size());
  for (size_t i = 0; i < lam_nm.size(); ++i) {
    g->SetPoint(i, lam_nm[i], T[i]);
  }

  // =========================
  // Rayleigh model
  // T = A * exp(-B / lambda^4)
  // =========================
  double xmin = *std::min_element(lam_nm.begin(), lam_nm.end())-10;
  double xmax = *std::max_element(lam_nm.begin(), lam_nm.end());

  TF1* f = new TF1("f", "[0]*exp(-[1]/pow(x,4))", xmin, xmax);
  f->SetParName(0, "A");
  f->SetParName(1, "B");
  f->SetParameters(0.95, 1e10);

  TFitResultPtr fitres = g->Fit(f, "S R");

  double A_fit = f->GetParameter(0);
  double B_fit = f->GetParameter(1);
  double A_err = f->GetParError(0);
  double B_err = f->GetParError(1);

  // =========================
  // Rayleigh coefficient C
  // =========================
  double C_nm4_per_mm = B_fit / t_mm;
  double C_nm4_per_mm_err = B_err / t_mm;

  // unit conversion: nm^4/mm -> um^4/cm
  double C_um4_per_cm = C_nm4_per_mm * 1e-11;
  double C_um4_per_cm_err = C_nm4_per_mm_err * 1e-11;

  // =========================
  // Transmittance at 400 nm
  // =========================
  double T400 = f->Eval(lam_eval_nm);

  // =========================
  // Transmission length A_T(400)
  // A_T = -d / ln T
  // =========================
  double AT400_cm = std::numeric_limits<double>::quiet_NaN();
  if (T400 > 0.0 && T400 < 1.0) {
    AT400_cm = -d_cm / std::log(T400);
  }

  // =========================
  // Print results
  // =========================
  std::cout << "==== Rayleigh scattering fit (paper-ready) ====" << std::endl;
  std::cout << "A = " << A_fit << " ± " << A_err << std::endl;
  std::cout << "C = (" << C_um4_per_cm << " ± " << C_um4_per_cm_err
            << ") um^4 / cm" << std::endl;
  std::cout << std::endl;
  std::cout << "T(" << lam_eval_nm << " nm) = " << T400
            << "  (" << T400 * 100.0 << " %)" << std::endl;
  std::cout << "A_T(" << lam_eval_nm << " nm) = " << AT400_cm << " cm" << std::endl;
  std::cout << "(using thickness d = " << d_cm << " cm = " << t_mm << " mm)" << std::endl;

  // =========================
  // Style
  // =========================
  gStyle->SetOptStat(0);
  gStyle->SetOptTitle(0);

  gStyle->SetTextFont(132);
  gStyle->SetLabelFont(132, "XYZ");
  gStyle->SetTitleFont(132, "XYZ");
  gStyle->SetLegendFont(132);

  gStyle->SetPadTickX(1);
  gStyle->SetPadTickY(1);

  // =========================
  // Plot
  // =========================
  TCanvas* c = new TCanvas("c", "Rayleigh fit", 800, 500);
  c->SetMargin(0.12, 0.05, 0.12, 0.05);

  g->SetMarkerStyle(20);
  g->SetMarkerSize(1.0);
  g->SetMarkerColor(kBlack);
  g->SetLineColor(kBlack);

  g->Draw("AP");
  g->GetXaxis()->SetTitle("Wavelength [nm]");
  g->GetYaxis()->SetTitle("Transmittance");
  g->GetXaxis()->SetTitleSize(0.05);
  g->GetYaxis()->SetTitleSize(0.05);
  g->GetXaxis()->SetLabelSize(0.045);
  g->GetYaxis()->SetLabelSize(0.045);

  g->GetYaxis()->SetRangeUser(0.0, 1.05);

  f->SetLineColor(kRed + 1);
  f->SetLineWidth(3);
  f->Draw("SAME");

  c->SaveAs("ag22k001_rayleigh_fit.pdf");
  //c->SaveAs("ag22k001_rayleigh_fit_root.png");
}
